### ISSUE TITLE
Some issues with PostProcessing

### ISSUE THREAD

**Issue by Sbozzolo** [ShaunFell/GRBoondi#13]
Here are some issues I see with the tools in PostProcessing:
- [x] There is no installation instruction
- [x] Dependencies are not declared anywhere, and some of the dependencies are not trivial (e.g., `numpy`, `pandas`, `matplotlib`)
- [x] There are no tests
- [x] It's unclear what versions of python are supported/required
- [x] It's unclear what versions of visit are supported/required
- [x] There is not much error handling (e.g., what if some output files already exist?, or what is `len(argv) < 1`?)
- [x] Scripts seem hard to understand/extend. There is not much explanation of what is happening, or what is the relationship between python and ini files.
- [x] Comments like the ones in lines https://github.com/ShaunFell/GRBoondi/blob/efa29d2ebc0b131aec2e761040dd9e133ed750c7/PostProcessing/IntegralsPlot.py#L156-L160 are not useful
- [x] Some tabs are 8 spaces, other 4
- [x] `SystemError` is typically used for problems like "disk full", or "file not found"

**Comment by ShaunFell** [comment]
Hi @Sbozzolo! I've made some changes per your suggestions in PR #18.
 * I added documentation on installing the right packages and using the post-processing routines on the wiki, [here](https://github.com/ShaunFell/GRBoondi/wiki/Visualization) and [here](https://github.com/ShaunFell/GRBoondi/wiki/Making-Plots)
 * I added more [error handling](https://github.com/ShaunFell/GRBoondi/pull/18/commits/9f14ecccac65312a48475189d2034db21d54a5d5) and general comments

As for the tests, I am still thinking of some good tests to setup, since the scripts are just fancy plotters for hdf5 data files. I suppose I could include a cheap HDF5 file and setup a test that produces a known plot, though thats not very elegant.

**Comment by Sbozzolo** [comment]
Hi @ShaunFell, thank you! Here are some follow up comments

> Since GRBoondi outputs the 3D data to HDF5 files, you can in theory use any visualization software you wish that has to capability to load HDF5 files. 

This is not really true. HDF5 is just a container, the way you structure your data inside the file matters a lot, so your visualization software has to know how to read the specific HDF5 files produced by your application. Visit knows how to read chombo files, but other programs might not know.

Some typos:
```diff
-Currently, there exists three plotting routines for data visualization, the first two leveraging Visit's extensive function list and the last using pythons Matplotlib
+Currently, there exists three plotting routines for data visualization, the first two leveraging Visit's extensive function list and the last using Matplotlib
```

```diff
-This is the flag specifying whether the i'th variables plot values should be clipped
+This is the flag specifying whether the i-th variables plot values should be clipped
```

```diff
-This is a list of 2 values specifying the lower and upper range of the i'th variables plot values
+This is a list of 2 values specifying the lower and upper range of the i-th variables plot values
```

```diff
-n the plot parameter file, you can specify a handleful of engine parameters, like launching the engine on a job scheduler, e.g. Slurm.
+n the plot parameter file, you can specify a handful of engine parameters, like launching the engine on a job scheduler, e.g. Slurm.
```

The [linked section for an explanation of the terminology](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/python_scripting/attributes.html#volume-volumeattributes) doesn't really explain the terminology

The "Line Plots" does not explain how to use `IntegralsPlots.py`

The `Making Plots` doesn't really say explain how to make plots and just lists the prerequisites for the "Visualization" page. It would be better to have something more robust than "I used this versions, anything later should work", if possible.

Can you include some examples of how the output from the various scripts looks like (especially for one of the examples)?

The minimum test that should be run is ensuring that the code executes successfully and produces a figure (ideally, in an automated way). It would be preferable to break your scripts down in smaller pieces that can be tested independently. 

The other comments in the first post are still relevant. For example:
https://github.com/ShaunFell/GRBoondi/blob/4e38914a435aa9b179a9f6261231609f8fb49283/PostProcessing/2DPlot.py#L12C1-L12C25
will lead to a failure without explanation if one calls the script without passing something. It is not immediately clear what needs to be passed. Most scripts provide an explanation on how to use them when they are called with the wrong number of arguments.
Or
https://github.com/ShaunFell/GRBoondi/blob/4e38914a435aa9b179a9f6261231609f8fb49283/PostProcessing/2DPlot.py#L370
there are no checks that some files actually exist.
Or 
https://github.com/ShaunFell/GRBoondi/blob/4e38914a435aa9b179a9f6261231609f8fb49283/PostProcessing/IntegralsPlot.py#L14-L15
what happens to preexisting output?
And many other examples like this

Plotting is one of the main ways users will interact with your code, so I think it is important to have robust scripts that users can easy understand and extend. You could split the logic in a small library + some executables so that people don't have to go thorugh 100s of lines to understand the basic flow.

**Comment by ShaunFell** [comment]
Hi @Sbozzolo! Thank you for your continued comments!

> This is not really true. HDF5 is just a container, the way you structure your data inside the file matters a lot, so your visualization software has to know how to read the specific HDF5 files produced by your application. Visit knows how to read chombo files, but other programs might not know.

Yes I agree that HDF5 is a container, but what I said is still true, since *in theory*, if the user knows how the container is structured, they can use any processing tool they desire. This would be useful for example for generating line plots of the 3D data. I have python in mind for this, since h5py can load the HDF5 files, then users can manipulate the data as they wish. Of course, VisIt is a much more powerful tool since it is already aware of the container structure. I will add a note below the paragraph you quoted on this.

Thanks for the typo corrections!

> The [linked section for an explanation of the terminology](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/python_scripting/attributes.html#volume-volumeattributes) doesn't really explain the terminology

Unfortunately, I am aware of this. I removed this link since it's probably not actually that helpful. VisIt doesnt document all of the attributes, so it was up to me to determine what they do. I wrote a complete [parameter guide](https://github.com/ShaunFell/GRBoondi/wiki/Making-Plots#plot-parameters-a-complete-guide) to explain all the parameters I assign in the scripts

> The Making Plots doesn't really say explain how to make plots and just lists the prerequisites for the "Visualization" page. It would be better to have something more robust than "I used this versions, anything later should work", if possible.

I completely overhauled this page and included examples as well as the parameter guide

> The minimum test that should be run is ensuring that the code executes successfully and produces a figure (ideally, in an automated way). It would be preferable to break your scripts down in smaller pieces that can be tested independently.

I built a new test suite for the post-processing routines. it downloads light-weight test data I uploaded to Zenodo and produces plots using all the plot routines and movies. To that end, I overhauled the PostProcessing directory and moved the functions/classes into a more package-like format in #18 . I also added more exception catching and error checking while I did this.

> Plotting is one of the main ways users will interact with your code, so I think it is important to have robust scripts that users can easy understand and extend. You could split the logic in a small library + some executables so that people don't have to go thorugh 100s of lines to understand the basic flow.

I agree. I suppose I originally designed it to be a bit of a black box that users could just run without worrying about whats inside, but thinking about it, that goes against the design philosophy of the core code of GRBoondi. The PostProcessing directory should be much more readable now.

**Comment by Sbozzolo** [comment]
@ShaunFell, thank you for the outstanding work in PR #18!

**Comment by ShaunFell** [comment]
Hi @Sbozzolo!

I've updated the PR. If youre satisfied with the changes, I'll go ahead and merge them into `main`.

**Comment by Sbozzolo** [comment]
I left some micro comments, but you addressed all my main concerns. Thank you for your work!

**Comment by ShaunFell** [comment]
Thanks for your review! I've merged PR #18
